### PR TITLE
fix the flaws in the test of the attach API

### PR DESCRIPTION
**Pull request by mountkin**
Signed-off-by: Shijiang Wei mountkin@gmail.com

### CODE REVIEW

File: integration-cli/docker_api_attach_test.go [hunk 1]
```diff
@@ -99,128 +100,81 @@ func (s *DockerSuite) TestGetContainersWsAttachContainerNotFound(c *check.C) {
...
-	outChan := make(chan error)
-	go func() {
-		_, err := r.Read(actual)
-		outChan <- err
-		close(outChan)
-	}()
-
-	inChan := make(chan error)
-	go func() {
-		_, err := w.Write(expected)
```
Comment (mountkin) [inline integration-cli/docker_api_attach_test.go]: This is just testing the builtin `io.Pipe()` functions.
Write to one end and read from the other end, while it did nothing to do with the attach API.

File: integration-cli/docker_api_attach_test.go [hunk 2]
```diff
@@ -99,128 +100,81 @@ func (s *DockerSuite) TestGetContainersWsAttachContainerNotFound(c *check.C) {
...
-	case err := <-outChan:
+	expectTimeout := func(conn net.Conn, br *bufio.Reader, stream string) {
+		defer conn.Close()
+		_, err := conn.Write([]byte{'t'})
 		c.Assert(err, check.IsNil)
-	case <-time.After(5 * time.Second):
-		c.Fatal("Timeout reading from stdin")
-	}
 
-	if !bytes.Equal(expected, actual) {
```
Comment (mountkin) [inline integration-cli/docker_api_attach_test.go]: As long as there are no bugs in the golang stdlib and the kernel this will almost always be true.

**Comment by mountkin** [comment]
@vdemeester Could you take a look at this PR before someone begin to refactor `integration-cli/docker_api_attach_test.go` per #16756? I deleted some of the incorrect `io.Pipe` logic in this PR.

BTW, I can finish the rest refactoring work in this PR but I'm not sure if it's better to open a separate PR to do the job.

**Comment by vdemeester** [comment]
@mountkin I've added you for this file in #16756 to make sure nobody takes this one :wink:.

I feel like there is not much less work to do to finish the rest of the refactoring, right ? I wouldn't mind having the whole file refactored in the same PR :wink:

**Comment by mountkin** [comment]
@vdemeester The whole file refactored. PTAL when you have time. Thanks! :smile:

**Comment by mountkin** [comment]
I mistakenly deleted my branch so this PR got closed. Reopening

File: integration-cli/docker_api_attach_test.go [hunk 3]
```diff
@@ -17,23 +18,17 @@ func (s *DockerSuite) TestGetContainersAttachWebsocket(c *check.C) {
 	out, _ := dockerCmd(c, "run", "-dit", "busybox", "cat")
 
 	rwc, err := sockConn(time.Duration(10 * time.Second))
-	if err != nil {
-		c.Fatal(err)
-	}
+	c.Assert(err, check.IsNil)
```
Comment (HuKeping) [inline integration-cli/docker_api_attach_test.go]: you can use `checker` now

File: integration-cli/docker_api_attach_test.go [hunk 4]
```diff
@@ -17,23 +18,17 @@ func (s *DockerSuite) TestGetContainersAttachWebsocket(c *check.C) {
...
 
 	cleanedContainerID := strings.TrimSpace(out)
 	config, err := websocket.NewConfig(
 		"/containers/"+cleanedContainerID+"/attach/ws?stream=1&stdin=1&stdout=1&stderr=1",
 		"http://localhost",
 	)
-	if err != nil {
-		c.Fatal(err)
-	}
+	c.Assert(err, check.IsNil)
```
Comment (HuKeping) [inline integration-cli/docker_api_attach_test.go]: same as above

File: integration-cli/docker_api_attach_test.go [hunk 5]
```diff
@@ -17,23 +18,17 @@ func (s *DockerSuite) TestGetContainersAttachWebsocket(c *check.C) {
...
 
 	ws, err := websocket.NewClient(config, rwc)
-	if err != nil {
-		c.Fatal(err)
-	}
+	c.Assert(err, check.IsNil)
```
Comment (HuKeping) [inline integration-cli/docker_api_attach_test.go]: here and below :)

**Comment by vdemeester** [comment]
Oh boy it's been long :sweat_smile: 

LGTM 🐶

ping @runcom @LK4D4

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**